enhancements/api-review/custom-tags-aws.md

```diff
@@ -0,0 +1,125 @@
+approvers:
+  - @decarr 
+creation-date: 2021-03-24
+last-updated: 2021-04-09
+status: implementable
+---
+
+# Apply user defined tags to all AWS resources created by OpenShift
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement describes a proposal to allow an administrator of OpenShift to
+have the ability to apply user defined tags to all resources created by OpenShift in AWS.
+
+Note: this enhancement is slightly retroactive. Work has already begun on this. See
+ - https://github.com/openshift/api/pull/864
+ - https://github.com/openshift/cluster-ingress-operator/pull/578
+
+## Motivation
+
+Motivations include but are not limited to:
+
+ - allow admin, compliance, and security teams to keep track of assets and objects created by OpenShift,
+   both at install time and during continuous operation (Day 2)
+ - in a Managed OpenShift environment such as Red Hat OpenShift on AWS (ROSA), allow easy differentiation
+   between Red Hat-managed AWS objects and customer-managed objects
+ - *allow for the restriction of permissions granted to Red Hat in an AWS account by AWS resource tags.
+   For example, see https://issues.redhat.com/browse/SDE-1146 - "IAM users and roles can only operate on resources with specific tags"*
+
+### Goals
+
+ - the administrator or service (in the case of Managed OpenShift) installing OpenShift can pass an arbitrary
+   list of user-defined tags to the OpenShift Installer, and everything created by the installer and all other
+   bootstrapped components will apply those tags to all resources created in AWS, for the life of the cluster.
+ - tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and,
+   after some period of time, apply the tags post-creation.
+
+### Non-Goals
+
+ - to reduce initial scope, tags are applied only at creation time and not reconciled. If an administrator manually
+   changes the tags stored in the infrastructure resource, behavior is undefined. See below.
+ - to reduce initial scope, we are not implementing this for clouds other than AWS. We will not take any actions
+   to prohibit that later.
+
+## Proposal
+
+New `experimentalPropagateUserTags` field added to `.platform.aws` of install config to indicate that the user tags should be applied to AWS
+resources created by in-cluster operators.
+
+If `experimentalPropagateUserTags` is true, install validation will fail if there is any tag that starts with `kubernetes.io` or `openshift.io`.
+
+Add a new field `userTags` to `.status.aws` of the `infrastructure.config.openshift.io` type. Tags included in the
```

Comment:
I disagree with putting these values in status.  For 1, it will make it more difficult for users to change if they need to for bug/support purposes.  For 2, if someone attempts to recreate this object from a backup or source control repo, oc apply will not add the elements to status.  Thus, this could create a situation where users have trouble restoring something after accidental deletion.

User defined values belong in spec.

Reply:
Is status not preserved on this field today with velero backup?  We have other write once status fields.

Reply:
etcd backup restore will preserve this as well.

Reply:
Looking at this issue, I don't think velero restores the status no: https://github.com/vmware-tanzu/velero/issues/1687 

I think in general there is an assumption that a resources status should be computable by the controller that reconciles it, though that doesn't apply when your CR is informational like this one is

Reply:
ok, it can be added to spec at a future time.  Present backup and restore is at etcd layer itself.  we can evolve this part post 4.8